src/Data/Aeson/BetterErrors/Internal.hs

```diff
@@ -42,28 +44,44 @@ import Data.Aeson.BetterErrors.Utils
 --     asTuple :: Parse e (Int, Int)
 --     asTuple = (,) \<$\> nth 0 asIntegral \<*\> nth 1 asIntegral
 -- @
-newtype Parse err a
-  = Parse (ReaderT ParseReader (Except (ParseError err)) a)
+newtype ParseT err m a
+  = ParseT (ReaderT ParseReader (ExceptT (ParseError err) m) a)
   deriving (Functor, Applicative, Monad,
             MonadReader ParseReader, MonadError (ParseError err))
+type Parse err a = ParseT err Identity a
+
+instance MonadTrans (ParseT err) where
+  lift f = ParseT (lift (lift f))
+
+runParseT :: ParseT err m a -> A.Value -> m (Either (ParseError err) a)
+runParseT (ParseT p) v = runExceptT (runReaderT p (ParseReader DList.empty v))
 
 runParse :: Parse err a -> A.Value -> Either (ParseError err) a
-runParse (Parse p) v = runExcept (runReaderT p (ParseReader DList.empty v))
+runParse p v = runIdentity (runParseT p v)
+
+mapParseT :: (ReaderT ParseReader (ExceptT (ParseError err) m) a -> ReaderT ParseReader (ExceptT (ParseError err') m') a') -> ParseT err m a -> ParseT err' m' a'
+mapParseT f (ParseT p) = ParseT (f p)
 
 -- | Transform the error of a parser according to the given function.
-mapError :: (err -> err') -> Parse err a -> Parse err' a
-mapError f p = do
-  v <- asks rdrValue
-  case runParse p v of
-    Right x -> pure x
-    Left err -> throwError (f <$> err)
+mapError :: Functor m => (err -> err') -> ParseT err m a -> ParseT err' m a
+mapError f = mapParseT (mapReaderT (withExceptT (fmap f)))
 
 -- | An infix version of 'mapError'.
-(.!) :: Parse err a -> (err -> err') -> Parse err' a
+(.!) :: Functor m => ParseT err m a -> (err -> err') -> ParseT err' m a
 (.!) = flip mapError
 
 -- | The type of parsers which never produce custom validation errors.
-type Parse' = Parse Void
+type Parse' a = Parse Void a
```

Comment:
Because `Parse` is now a type synonym rather than a newtype, ghc doesn't let you partially apply it by default without enabling some extension. It means the same thing, otherwise.